llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp

```diff
@@ -16,26 +16,31 @@
 
 using namespace llvm;
 
-std::map<StringRef, std::vector<uint32_t>>
-llvm::getSYCLDeviceRequirements(const Module &M) {
-  std::map<StringRef, std::vector<uint32_t>> Result;
+void llvm::getSYCLDeviceRequirements(
+    const Module &M, std::map<StringRef, std::vector<uint32_t>> &Requirements) {
   auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
                                              unsigned OpNo) -> unsigned {
     Constant *C =
         cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
     return static_cast<uint32_t>(C->getUniqueInteger().getZExtValue());
   };
-  std::set<uint32_t> Aspects;
-  for (const Function &F : M) {
-    if (!F.hasMetadata("sycl_used_aspects"))
-      continue;
 
-    const MDNode *MD = F.getMetadata("sycl_used_aspects");
-    for (size_t I = 0, E = MD->getNumOperands(); I < E; ++I) {
-      Aspects.insert(ExtractIntegerFromMDNodeOperand(MD, I));
+  std::vector<std::pair<std::string, const char *>> ReqdMDs = {
+      {"sycl_used_aspects", "aspects"}, {"sycl_fixed_targets", "fixed_target"}};
```

Comment:
```suggestion
  constexpr std::pair<const char *, const char *> ReqdMDs[] = {
      {"sycl_used_aspects", "aspects"}, {"sycl_fixed_targets", "fixed_target"}};
```
That should give the compiler an easier time with optimizing the loop. Not though that it cannot have an `std::string` in there, so you may want to do a conversion or a C-style string comparison later in the code. The comparison might just work since they will both be the same string literal, but better be on the safe side.

Reply:
Thanks, and yes, we don't need a vector here. [8c910f7](https://github.com/intel/llvm/pull/7095/commits/8c910f72197ba055b1a2bfb3dcc2a8b2cb3d23e5)